chainer/optimizer.py

```diff
@@ -185,11 +185,31 @@ def update(self, param):
             return
 
         self.t += 1
-        if param.data is not None:
-            self._prepare(param)
-        for hook in six.itervalues(self._hooks):
-            hook(self, param)
-        self.update_core(param)
+
+        use_fp32_update = getattr(self, "_use_fp32_update", False)
```

Comment:
Adding the default values of `_use_fp32_update` and `fp32_param` as class fields (or instance fields initialized in `__init__`) is simpler:
```py
_use_fp32_update = False
_fp32_param = None

def update(self, param):
    ...
    if self._use_fp32_update and param.dtype == numpy.float16:
        if self._fp32_param is None:
            ...
```